[JPMS Support] Refactor the codebase to eliminate top level split packages

prudhvigodithi opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description

- For Phase 0 - JPMS Support, refactor `:server` module `cli` package to eliminate top level split packages.

- The `:libs:opensearch-cli` and `:server` module both has `org.opensearch.cli`. Updated the `:server` `org.opensearch.cli` to `org.opensearch.common.cli` (part of `:server`).

- Deleted the class `server/src/main/java/org/opensearch/cli/LoggingAwareCommand.java` which is not used.

- Moved the common existing cli tests from `:server` to `:libs` (`libs/cli/src/test/java/cli/`).

- This change requires a simple refactor to eliminate the top level split package issue for `org.opensearch.cli`.

- Fix `:distribution:tools:keystore-cli` split level package problem `org.opensearch.common.settings`. Modify to  `org.opensearch.cli.keystore`.

- Removed the `libs/plugin-classloader` and refactored the `ExtendedPluginsClassLoader` to  `:server`.

- Moved some of the `InstallPluginCommand` and `RemovePluginCommand` logic to `PluginsService` itself, this way the `Bundle`, `plugin` and some of its methods that instantiates `Bundle` can be private (from `PluginsService`). 

### Related Issues
Part of https://github.com/opensearch-project/OpenSearch/issues/8110

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

prudhvigodithi commented:
Adding @msfroh @reta @andrross to please take a look and your thoughts.
Thanks 
@getsaurabh02

msfroh commented:
Looking through a lot of these CLI commands, I think we should move them out of `:server`. It's a tricky dance, though.

My intuition is as follows:

1. `KeyStoreAwareCommand` (and its descendants) should *probably* move to `:distribution:tools:keystore-cli` (though we need to move the constant `MAX_PASSPHRASE_LENGTH` to some common package, unless we're okay with duplicating it).
2. Ideally, we'd move `EnvironmentAwareCommand` into `:lib:cli`, but it depends on `Environment` and `Settings`, which are both in `:server`.

That's where it becomes clear that we have a lot more detangling to do (as highlighted on the linked issue). Can we at least try to move `KeyStoreAwareCommand` and friends to `keystore-cli`? (Even if it means moving `MAX_PASSPHRASE_LENGTH` somewhere else.)

prudhvigodithi commented:
> That's where it becomes clear that we have a lot more detangling to do (as highlighted on the linked issue). Can we at least try to move `KeyStoreAwareCommand` and friends to `keystore-cli`? (Even if it means moving `MAX_PASSPHRASE_LENGTH` somewhere else.)

When we try to pull the `KeyStoreAwareCommand` and its friends out of `:server` it becomes an bigger mess, we need to pull `org.opensearch.node.`,  `org.opensearch.env.`,  `org.opensearch.common.settings`,  `org.opensearch.common.logging` out of server, which again has its transitive dependencies example `org.opensearch.client` which are again part of the `:server`. The high level goal of this PR is to support JPMS by eliminating top level split packages.

prudhvigodithi commented:
Let me explore and see the refactoring (moving) to `keystore-cli` and I will add some of analysis here.

prudhvigodithi commented:
@msfroh while working on moving `KeyStoreAwareCommand` and its related similar classes to `keystore-cli`,  I have noticed the `:distribution:tools:keystore-cli` has a split level package problem `org.opensearch.common.settings`, hence I have refactored to  `org.opensearch.cli.keystore`.  I was able to move the keystore related classes to `:distribution:tools:keystore-cli` and fixed the problem related to `MAX_PASSPHRASE_LENGTH`, onboarded `KeystoreConstants` in `:libs:opensearch-cli` `org.opensearch.cli` which is common can be shared by the other modules. Please check.
@reta @andrross

prudhvigodithi commented:
Added missing Javadoc for`:distribution:tools:keystore-cli:` to the best of my knowledge (seen some assemble failures for `missingJavadoc`).

reta commented on libs/cli/src/test/java/org/opensearch/cli/KeystoreConstantsTests.java:
```diff
@@ -0,0 +1,19 @@
+
+import org.opensearch.test.OpenSearchTestCase;
+
+public class KeystoreConstantsTests extends OpenSearchTestCase {
```
This tests is not very useful to be fair

prudhvigodithi replied:
Initially I just added as placeholder but since with the move of `MAX_PASSPHRASE_LENGTH ` to `EnvironmentAwareCommand` removed this.

reta commented on server/src/main/java/org/opensearch/bootstrap/Bootstrap.java:
```diff
@@ -263,7 +263,7 @@ static SecureSettings loadSecureSettings(Environment initialEnv) throws Bootstra
         try {
             if (keystore != null && keystore.hasPassword()) {
-                password = readPassphrase(System.in, KeyStoreAwareCommand.MAX_PASSPHRASE_LENGTH);
+                password = readPassphrase(System.in, KeystoreConstants.MAX_PASSPHRASE_LENGTH);
```
This is the only place `MAX_PASSPHRASE_LENGTH` is used, right? If yes, let's move it here

msfroh replied:
It's also used in the `KeyStoreAwareCommand#readPassword` method:

```
            passwordArray = terminal.readSecret(
                "Enter new password for the opensearch keystore (empty for no password): ",
                MAX_PASSPHRASE_LENGTH
            );
            char[] passwordVerification = terminal.readSecret("Enter same password again: ", MAX_PASSPHRASE_LENGTH);
```

reta commented on libs/cli/src/main/java/org/opensearch/cli/KeystoreConstants.java:
```diff
@@ -0,0 +1,15 @@
+
+package org.opensearch.cli;
+
+public class KeystoreConstants {
```
I don't think we need it, and even if we need it - shouldn't be part of KeyStoreWrapper or Keystore CLI?

msfroh replied:
Moving it somewhere "neutral" was my suggestion above. 

It's used by both the Keystore CLI (as part of `KeyStoreAwareCommand`) and by the server (in the `Bootstrap` class).

reta replied:
Oh I see, may be move this constant to `EnvironmentAwareCommand` ? Not really the ideal, but `KeystoreConstants` does not belong here (I think)

prudhvigodithi replied:
Added to `EnvironmentAwareCommand`. Removed `KeystoreConstants` and `KeystoreConstantsTests`.

reta commented on server/src/main/java/org/opensearch/common/cli/package-info.java:
```diff
@@ -9,4 +9,4 @@
  * The command line interface module.
  */
-package org.opensearch.cli;
+package org.opensearch.common.cli;
```
Just a question @msfroh  @prudhvigodithi what do you think folks the right name for the package  should be?

So we have:

```
org.opensearch.cli
org.opensearch.cli.keystore
```

It looks to me we should have:

```
org.opensearch.cli.common
```

What do you think?

andrross replied:
@reta It makes sense that all CLI classes use the `o.o.cli` prefix. I just don't like the `:server` module would "own" the generic sounding `o.o.cli.common` namespace. How about using `o.o.cli.server` since these are the CLI-related classes that have a dependency on `:server` components?

prudhvigodithi replied:
We should go with `org.opensearch.common.cli`, as it indicates the location of the common CLI classes for the `:server` module (from other packages inside :server common).

andrross replied:
I would personally like to eventually get rid of any usage of `o.o.common` in the `:server` module, as I would expect `o.o.common` to be used by `:libs:opensearch-common`, so that makes `org.opensearch.common.cli` my least preferred option.

andrross commented on distribution/tools/keystore-cli/src/main/java/org/opensearch/cli/keystore/package-info.java:
```diff
@@ -9,4 +9,4 @@
  * Classes implementing a CLI tool for managing secrets in the OpenSearch keystore.
  */
-package org.opensearch.common.settings;
+package org.opensearch.cli.keystore;
```
How about making this `org.opensearch.tools.cli.keystore`? The `java-version-checker` and `launcher` modules within this "tools" folder make use of the `org.opensearch.tools` prefix. Seems like a good opportunity to make things consistent and also make the same change to `plugin-cli` and `upgrade-cli`. I don't feel super strongly though.

prudhvigodithi replied:
Yes we should do that and keep it consistent.

msfroh commented on server/src/main/java/org/opensearch/common/settings/KeyStoreWrapper.java:
```diff
@@ -134,7 +134,7 @@ private static class Entry {
 
     /** The version of the metadata written before the keystore data. */
-    static final int FORMAT_VERSION = 4;
+    public static final int FORMAT_VERSION = 4;
```
Instead of changing the visibility of these members, can we move `KeyStoreWrapper` into the `org.opensearch.cli.keystore` package?

prudhvigodithi replied:
Ya initially I have tried that moving `KeyStoreWrapper` to keystore-cli and for server gradle file add a line `  api project(":distribution:tools:keystore-cli")`, but this breaks with.
```
Circular dependency between the following tasks:
:server:compileJava
\--- :distribution:tools:keystore-cli:compileJava
     +--- :server:compileJava (*)
     \--- :distribution:tools:keystore-cli:compileJava (*)

(*) - details omitted (listed previously)
```
Also I assume the KeyStoreWrapper is a wrapper for `org.opensearch.cli.keystore` from `:server ` `org.opensearch.common.settings`.

prudhvigodithi replied:
Also I see a mix, example the `SEED_SETTING` is also public `public static final Setting<SecureString> SEED_SETTING = SecureSetting.secureString("keystore.seed", null);`

prudhvigodithi commented:
- Refactored `keystore-cli` to `org.opensearch.tools.cli.keystore`, `plugin-cli` to `org.opensearch.tools.cli.plugin`,  `upgrade-cli`  to `org.opensearch.tools.cli.upgrade` and updated the `:qa:evil-tests` accordingly.

- Refactored the `:libs:plugin-classholder` to `org.opensearch.plugin_classloader`.

msfroh reviewed (approved):
I think this is about as good as we're going to get without a lot more refactoring, and is IMO a great start. Thanks @prudhvigodithi !

@reta , @andrross -- do you have more feedback?

reta commented on distribution/tools/keystore-cli/src/main/java/org/opensearch/tools/cli/keystore/UpgradeKeyStoreCommand.java:
```diff
@@ -43,10 +44,26 @@
+     * Initializes a command that requires an existing keystore to upgrade its format
+     * to the latest version. This command will fail if the keystore doesn't exist.
+     */
+    public UpgradeKeyStoreCommand() {
```
Why it is public? 

```suggestion
    UpgradeKeyStoreCommand() {
```

prudhvigodithi replied:
Fixed this.

reta commented on server/build.gradle:
```diff
@@ -70,6 +70,7 @@ dependencies {
   api project(":libs:opensearch-telemetry")
   api project(":libs:opensearch-task-commons")
 
+
```
Accidental change?

prudhvigodithi replied:
Ya reverted back, coming from https://github.com/opensearch-project/OpenSearch/pull/17153#discussion_r1933007148 was playing with some gradle dependencies.

reta commented on server/src/main/java/org/opensearch/plugins/PluginsService.java:
```diff
@@ -321,12 +321,19 @@ public PluginsAndModules info() {
     // a "bundle" is a group of jars in a single classloader
-    static class Bundle {
-        final PluginInfo plugin;
+    public static class Bundle {
```
Please bring back the package protected visibility, these classes where not designed as public

prudhvigodithi replied:
Ya PluginsService.java the methods and class is used by by both `:server` and `:plugin-cli`, with this refactoring some of them had to be changed to public.

reta replied:
You are right, getPluginBundles exposes the Bundle, but we don't need to expose constructors and `plugin`, this is for sure

prudhvigodithi replied:
Let me explore few ways here.

reta commented on server/build.gradle:
```diff
@@ -69,7 +69,7 @@ dependencies {
   api project(":libs:opensearch-telemetry")
   api project(":libs:opensearch-task-commons")
-
+  
```
Accidental?

reta commented on libs/plugin-classloader/src/main/java/org/opensearch/plugin_classloader/ExtendedPluginsClassLoader.java:
```diff
@@ -30,7 +30,7 @@
  */
 
-package org.opensearch.plugins;
+package org.opensearch.plugin_classloader;
```
```suggestion
package org.opensearch.plugins.classloader;
```

prudhvigodithi replied:
Sure but I just followed this https://github.com/opensearch-project/OpenSearch/tree/main/libs/secure-sm/src/main/java/org/opensearch `org.opensearch.secure_sm`

prudhvigodithi replied:
Should we go with `org.opensearch.plugin_classloader;` or `org.opensearch.plugins.classloader;` ?

reta replied:
I think we had `plugins` before, right?

prudhvigodithi replied:
Ya we have already one `org.opensearch.plugins` inside `:server`.

prudhvigodithi replied:
Looking more closely we can move the `ExtendedPluginsClassLoader` to `:server` plugins, we dont need a separate module for this.

prudhvigodithi replied:
I have removed the `libs/plugin-classloader` and refactored to `:server`. @reta please check.

reta replied:
> I have removed the `libs/plugin-classloader` and refactored to `:server`. @reta please check.

I suspect it will backfire, this module was used as security boundary to contain the `createClassloader` calls, you may need to bring it back I guess

prudhvigodithi replied:
True it failed lot of stuff, for now I will keep the `libs/plugin-classloader`.

prudhvigodithi replied:
@reta should we go with `org.opensearch.plugin.classloader`, looks to me like this not create split level package problem.

reta commented on distribution/build.gradle:
```diff
@@ -481,7 +481,7 @@ configure(subprojects.findAll { ['archives', 'packages'].contains(it.name) }) {
       }
     }
-    
+
```
Accidental?

prudhvigodithi replied:
Looks like this is coming from `./gradlew spotlessApply ` when I ran on local.
